KAFKA-3316: Add REST API for listing connector plugins

Ishiihara opened the pull request.

Ishiihara committed:
add list connector plugins

Ishiihara commented:
@ewencp @gwenshap Ping for review.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java:
```diff
@@ -189,6 +190,16 @@ public ConfigInfos validateConfigs(String connType, Map<String, String> connecto
+    @Override
+    public List<String> connectorPlugins() {
+        List<String> connectorPlugins = new LinkedList<>();
+        Set<Class<? extends Connector>> connectorClasses = Worker.getConnectorPlugins();
```
You can just do something like `return new ArrayList<String>(Worker.getConnectorPlugins())`. No need to manually copy with a for loop.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -77,6 +81,8 @@
     private KafkaProducer<byte[], byte[]> producer;
     private SourceTaskOffsetCommitter sourceTaskOffsetCommitter;
+    private static final List<Class<? extends Connector>>
+        SKIPPED_CONNECTORS = Arrays.asList(SourceConnector.class, SinkConnector.class, VerifiableSourceConnector.class, VerifiableSinkConnector.class);
```
Special casing some of these makes sense (e.g. the `Verifiable*Connector` classes), but others are skipped for a different reason: they are interfaces/abstract classes. Can we handle these more generally by checking the `Class.getModifiers()` for each class maybe? Ideally we could filter out abstract classes added by connector developers as well (which they might have if they can share code between a few different connectors).

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -194,6 +200,13 @@ public Connector getConnector(String connType) {
         return instantiateConnector(connectorClass);
     }
 
+    public static Set<Class<? extends Connector>> getConnectorPlugins() {
```
Is there any reason this needs to be in `Worker`? Seems like we could just do all of this in `AbstractHerder` and avoid introducing this extra API in `Worker`?

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -194,6 +200,13 @@ public Connector getConnector(String connType) {
     }
 
+    public static Set<Class<? extends Connector>> getConnectorPlugins() {
+        Reflections reflections = new Reflections(new ConfigurationBuilder().setUrls(ClasspathHelper.forJavaClassPath()));
```
Scanning for plugins is actually quite slow. Should we cache the results after the first lookup? I'm not sure we need to worry about anything like dynamically loaded classes, so I think that would be a safe way to avoid the cost of this scan for every request.

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorPluginsResourceTest.java:
```diff
@@ -120,6 +128,31 @@ public ConfigInfos answer() {
+    public void testListConnectorPlugins() {
+        herder.connectorPlugins();
+
+        PowerMock.expectLastCall().andAnswer(new IAnswer<List<String>>() {
```
Since this is basically just a pass-through method anyway, there isn't that much to test here -- you could simplify this test just to use a simple set of connectors you create yourself. There's a lot of code in this test when all you really want to see is that the set makes it back out of the call to ConnectorPluginsResource

Ishiihara committed:
change listConnectors to static method

Ishiihara commented:
@ewencp Addressed review comments. The reason for changing `listConnectors` to static is that the method does not depend on any state of `AbstractHerder`. Should be good to take another look.

Ishiihara committed:
Return a list of ConnectorPluginInfo

ewencp commented:
LGTM, thanks!
